meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java

```diff
@@ -0,0 +1,125 @@
+ * This LeaderElection client will notify registered listeners for any leadership change.
+ */
+public class LeaderElectionClient {
+
+  /**
+   * Construct a LeaderElectionClient using a user passed in leaderElectionConfig. It creates a MetaClient
+   * instance underneath.
+   * When MetaClient is auto closed be cause of being disconnected and auto retry connection timed out, A new
+   * MetaClient instance will be created and keeps retry connection.
+   */
+  public LeaderElectionClient(MetaClientConfig metaClientConfig) {
+
+  }
+
+  /**
+   * Construct a LeaderElectionClient using a user passed in MetaClient object
+   * When MetaClient is auto closed be cause of being disconnected and auto retry connection timed out, user
+   * will need to create a new MetaClient and a new LeaderElectionClient instance.
+   */
+  public LeaderElectionClient(MetaClientInterface MetaClient) {
+
+  }
+
+  /**
+   * Returns true if current participant is able to acquire leadership.
+   */
+  public boolean isLeader(String leaderPath, String participant) {
+    return false;
+  }
+
+  /**
+   * Participants join a leader election group by calling the following API.
+   * The Leader Election client maintains and elect an active leader from participant pool
+   */
+  public boolean joinLeaderElectionParticipantPool(String leaderPath, String participant,
+      Object userInfo) {
+    return false;
+  }
+
+  /**
+   * Any participant may exit the exitLeaderElectionParticipantPool by calling the API.
+   * If the participant is not the current leader, it leaves the pool and won't be elected as next leader.
+   * If the participant is the current leader, it leaves the pool and a new leader will be elected if there are other participants in the pool.
+   * Throws exception if the participant is not in the pool.
+   */
+  public boolean exitLeaderElectionParticipantPool(String leaderPath, String participant) {
+    return false;
+  }
+
+  /**
+   * Releases leadership for participant.
+   * Throws exception if the leadership is not owned by this participant.
+   */
+  public void relinquishLeader(String leaderPath, String participant) {
+  }
+
+  /**
+   * Get current leader nodes.
+   * Returns null if no Leader at a given point.
+   */
+  public String getLeader(String leaderPath) {
+    return null;
+  }
+
+  /**
+   * Return a list of hosts in participant pool
+   */
+  public List<String> getFollowers(String leaderPath) {
+    return null;
+  }
+
+  /**
+   * APIs to register/unregister listener to leader path. All the participants can listen to any
+   * leaderPath, Including leader going down or a new leader comes up.
+   * Whenever current leader for that leaderPath goes down (considering it's ephemeral entity which
+   * get's auto-deleted after TTL or session timeout) or a new leader comes up, it notifies all
+   * participants who have been listening on entryChange event.
+   */
+  public boolean subscribeLeadershipChanges(String leaderPath,
```

Comment:
what is the contract - when leaderPath doesn't exist? will we create and wait for others to participate?

Reply:
subscribe for change is not a blocking call. If there is no leaderPath we still allow user to register listener and will notify if any change.